modules/oe_multilingual_url_suffix/src/Form/NegotiationUrlSuffixForm.php

```diff
@@ -0,0 +1,139 @@
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function getEditableConfigNames() {
+    return ['oe_multilingual_url_suffix.settings'];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildForm(array $form, FormStateInterface $form_state) {
+    global $base_url;
+    $config = $this->config('oe_multilingual_url_suffix.settings');
+
+    $form['suffix'] = [
+      '#type' => 'details',
+      '#tree' => TRUE,
+      '#title' => $this->t('Path suffix configuration'),
+      '#open' => TRUE,
+      '#description' => $this->t('Language codes or other custom text to use as a path suffix for URL language detection. For the selected fallback language, this value may be left blank. <strong>Modifying this value may break existing URLs. Use with caution in a production environment.</strong> Example: Specifying "deutsch" as the path suffix code for German results in URLs like "example.com/contact_deutsch".'),
+    ];
+
+    $languages = $this->languageManager->getLanguages();
+    $suffixes = $config->get('url_suffixes');
+    foreach ($languages as $langcode => $language) {
+      $t_args = ['%language' => $language->getName(), '%langcode' => $language->getId()];
+      $form['suffix'][$langcode] = [
+        '#type' => 'textfield',
+        '#title' => $language->isDefault() ? $this->t('%language (%langcode) path suffix (Default language)', $t_args) : $this->t('%language (%langcode) path suffix', $t_args),
+        '#maxlength' => 64,
+        '#default_value' => isset($suffixes[$langcode]) ? $suffixes[$langcode] : substr($langcode, 0, 2),
+        '#field_prefix' => $base_url . '/index' . LanguageNegotiationUrlSuffix::SUFFIX_DELIMITER,
+      ];
+    }
+
+    $form_state->setRedirect('language.negotiation');
+
+    return parent::buildForm($form, $form_state);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validateForm(array &$form, FormStateInterface $form_state) {
+    $languages = $this->languageManager->getLanguages();
+
+    // Count repeated values for uniqueness check.
+    $count = array_count_values($form_state->getValue('suffix'));
+    foreach ($languages as $langcode => $language) {
+      $value = $form_state->getValue(['suffix', $langcode]);
+      if ($value === '') {
+        // Throw a form error if the suffix is blank for any language,
+        // although it is required for selected negotiation type.
+        $form_state->setErrorByName("suffix][$langcode", $this->t('The suffix may only be left blank for the <a href=":url">selected detection fallback language.</a>', [
+          ':url' => $this->getUrlGenerator()->generate('language.negotiation_selected'),
+        ]));
+      }
+      elseif (strpos($value, LanguageNegotiationUrlSuffix::SUFFIX_DELIMITER) !== FALSE) {
+        // Throw a form error if the string contains an underscore,
+        // which would not work.
+        $form_state->setErrorByName("suffix][$langcode", $this->t('The suffix may not contain a underscore.'));
```

Comment:
Let's use the `LanguageNegotiationUrlSuffix::SUFFIX_DELIMITER` constant, instead of `underscore`. Something like:
```php
$this->t('The suffix may not contain the delimiter character: ""@delimiter.', [...]));
```

Reply:
Changed